File: solutions/python/08/branch_bound/main.py
def subset_sum_pruned(xs, target):
    xs = sorted(xs, reverse=True)
    best = [False]

    def dfs(i, rem):
        if rem == 0:
            best[0] = True
            return
        if rem < 0 or i == len(xs):
            return
        if sum(xs[i:]) < rem:
            return
        dfs(i + 1, rem - xs[i])
        if best[0]:
            return
        dfs(i + 1, rem)

    dfs(0, target)
    return best[0]


def main() -> None:
    xs = [3, 7, 8, 2, 5]
    for t in [10, 11, 23, 24]:
        print(t, subset_sum_pruned(xs, t))


if __name__ == "__main__":
    main()

File: solutions/go/08/branch_bound/main.go
package main

import "fmt"

func subsetSumPruned(xs []int, target int) bool {
	sortDesc := make([]int, len(xs))
	copy(sortDesc, xs)
	for i := 0; i < len(sortDesc); i++ {
		for j := i + 1; j < len(sortDesc); j++ {
			if sortDesc[j] > sortDesc[i] {
				sortDesc[i], sortDesc[j] = sortDesc[j], sortDesc[i]
			}
		}
	}
	best := false
	var dfs func(i, rem int)
	dfs = func(i, rem int) {
		if rem == 0 {
			best = true
			return
		}
		if rem < 0 || i == len(sortDesc) {
			return
		}
		s := 0
		for j := i; j < len(sortDesc); j++ {
			s += sortDesc[j]
		}
		if s < rem {
			return
		}
		dfs(i+1, rem-sortDesc[i])
		if best {
			return
		}
		dfs(i+1, rem)
	}
	dfs(0, target)
	return best
}

func main() {
	xs := []int{3, 7, 8, 2, 5}
	for _, t := range []int{10, 11, 23, 24} {
		fmt.Println(t, subsetSumPruned(xs, t))
	}
}
